python/az/aro/azext_aro/custom.py

```diff
@@ -128,9 +135,56 @@ def aro_create(cmd,  # pylint: disable=too-many-locals
                        parameters=oc)
 
 
-def aro_delete(client, resource_group_name, resource_name, no_wait=False):
+def aro_delete(cmd, client, resource_group_name, resource_name, no_wait=False):
     # TODO: clean up rbac
 
+    oc = None
+    try:
+        oc = client.get(resource_group_name, resource_name)
+    except CloudError:
+        pass
+
+    if oc:
+        master_subnet = oc.master_profile.subnet_id
+        worker_subnets = {w.subnet_id for w in oc.worker_profiles}
+
+        master_parts = parse_resource_id(master_subnet)
+        vnet = resource_id(
+            subscription=master_parts['subscription'],
+            resource_group=master_parts['resource_group'],
+            namespace='Microsoft.Network',
+            type='virtualNetworks',
+            name=master_parts['name'],
+        )
+
+        aad = AADManager(cmd.cli_ctx)
+
+        sp_ids = []
+
+        client_id = oc.service_principal_profile.client_id
+        client_sp = aad.get_service_principal(client_id)
+        if client_sp:
+            sp_ids.append(client_sp.object_id)
+        else:
+            logger.warning(
+                'Unable to retrieve the cluster service principal. This means '
+                'that it may have been deleted, and some cleanup may fail.'
+            )
+
+        rp_client_id = os.environ.get('AZURE_FP_CLIENT_ID', FP_CLIENT_ID)
+
+        rp_client_sp = aad.get_service_principal(rp_client_id)
+        if rp_client_sp:
+            sp_ids.append(rp_client_sp.object_id)
```

Comment:
So @jim-minter I went to implement your suggested changes, but as far as I can tell, the code as written here does not introduce a regression that would prevent deletion if there's some issue with the RP SPN? In both the client and RP SPN cases, if we can't get them, we don't append them to the list to tweak perms for.

Is the issue that we may be able to retrieve the SPN but its permissions would be reduced?

Do we just want to `try:` the block below and pass if this fails, similar to the case wherein we can't get the cluster to try this?

Reply:
@ehashman if the caller doesn't have permissions to call get_service_principal (example: the caller is a service principal with minimal AAD permissions), get_service_principal will raise an exception and the delete will not be called.  In our environment, $AZURE_RP_CLIENT_ID is such an SP.